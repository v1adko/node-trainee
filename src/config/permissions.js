export default {
  PUBLIC: { role: 'public', value: 0 },
  USER: { role: 'user', value: 1 },
  ADMIN: { role: 'admin', value: 2 }
};
