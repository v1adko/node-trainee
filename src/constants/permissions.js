export default {
  PUBLIC: { value: 'public', priority: 0 },
  USER: { value: 'user', priority: 1 },
  MODERATOR: { value: 'moderator', priority: 2 },
  ADMIN: { value: 'admin', priority: 3 }
};
