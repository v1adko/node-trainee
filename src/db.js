const { flags, connectionDBString } = require('./config');
const mongoose = require('mongoose');

mongoose.Promise = global.Promise;
if (flags.debug) {
  mongoose.set('debug', true);
}

let db = null;

const connect = () => {
  if (db === null) {
    db = mongoose.connect(connectionDBString, {
      useMongoClient: true
    });
  } else {
    throw new Error('DB instance already exists, use existing connection or close it before creating a new one.');
  }
};

const getConnection = () => {
  if (db) {
    return db;
  }
  throw new Error("DB connection doesn't exist yet.");
};

const close = () => {
  if (db) {
    db.close(() => {
      console.log('Mongoose disconnected on app');
    });
  } else {
    throw new Error("DB connection doesn't exist yet.");
  }
};

module.exports = {
  connect,
  getConnection,
  close
};

