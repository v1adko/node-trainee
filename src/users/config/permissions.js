export default {
  CREATE_USER: 'createUser',
  READ_USER: 'readUser',
  UPDATE_USER: 'updateUser',
  DELETE_USER: 'deleteUser',
  GET_MY_DATE: 'getMyDate',
  UPDATE_MY_DATE: 'updateMyDate'
};
