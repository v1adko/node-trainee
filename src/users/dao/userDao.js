const BaseDao = require('./baseDaoMongoose');
const User = require('../models/user');

class UserDao extends BaseDao {}

module.exports = new UserDao(User);
