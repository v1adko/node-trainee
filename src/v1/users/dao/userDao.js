import BaseDao from '../../../utils/baseDaoMongoose';
import User from '../models/user';

class UserDao extends BaseDao {
  create(username, password) {
    const user = new this.Model();
    user.username = username;
    user.password = password;
    return super.create(user);
  }
}

export default new UserDao(User);
