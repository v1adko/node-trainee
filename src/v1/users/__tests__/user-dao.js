import userDao from '../user-dao';
import userFields from '../../../utils/tests-utils/test-user-fields';
import db from '../../../db';
import permissions from '../../../constants/permissions';

const role = permissions.ADMIN.value;
const { username, password } = userFields;
const clean = async () => {
  await userDao.Model.remove({});
};

beforeAll(db.connect);
afterAll(db.closeConnection);

describe('Test base dao for mongoose', async () => {
  afterEach(clean);

  it('shold add test object to database', async () => {
    const user = await userDao.create(username, password, role);

    expect(user.username).toBe(username);
    expect(user.role).toBe(role);
  });

  it('shold add test object to database, witout role', async () => {
    const user = await userDao.create(username, password);

    expect(user.username).toBe(username);
    expect(user.role).toBe(permissions.USER.value);
  });

  it('shold not add test object with same unique fields', async () => {
    await userDao.create(username, password);

    let duplicateError = null;
    try {
      await userDao.create(username, password);
    } catch (error) {
      duplicateError = error;
    }

    expect(duplicateError).toMatchSnapshot();
  });
});
