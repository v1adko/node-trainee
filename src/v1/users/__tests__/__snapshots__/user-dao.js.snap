// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test base dao for mongoose shold not add test object with same unique fields 1`] = `[ResourceDuplicateError: Cannot create resource, it has used unique fields what was engaged. User already exist. Please input another username.]`;
